FIXES: 7971 [2025-06 LWG Motion 34] P3552R3 Add a Coroutine Task

dietmarkuehl opened the pull request:
I have applied the changes for [P3552r3](https://wiki.edg.com/pub/Wg21sofia2025/StrawPolls/P3552R3.html) (Add a Coroutine Task Type). Compared to the actual text of [P3552r3](https://wiki.edg.com/pub/Wg21sofia2025/StrawPolls/P3552R3.html)
I made some [I think editorial] changes:

- added with_error and change_coroutine_scheduler declarations to the synopsis before task; they are in the detailed description but not in the synopsis (possibly they should only be in the synopsis) as there is no further description)
- inconsistent default template arguments for the ctor in the class declaration (allocator<byte> and the later description (allocator<void>); the wording consistently uses allocator<void>
- added "of the" to a sentence about the type of a completion datum: "... defines the type *of the* value completion datum"
- the use of the template parameter of the nested type state is Rcvr which has two problems: 1. the ctor also uses the name Rcvr for its template parameter 2. the class stores an object of type R which is the receiver named rcvr the most logic fix is to use Rcvr consistently for the state and change the name for the ctor's parameter:
    1. change the ctor template parameter type to use R
    2. change the member to use Rcvr
    3. change the type R used for the own-env-t to be "...get_env(declval<Rcvr>())..."

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+requirements\iref{allocator.requirements.general}.
+
+\rSec3[task.members]{\tcode{task} Members}
+
```
```suggestion

\indexlibraryctor{task}%
```

Various other index entries missing in this area.

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+
+\begin{itemdecl}
+template<@\libconcept{receiver}@ R>
+@\exposid{state}@<R> connect(R&& recv);
```
```suggestion
  @\exposid{state}@<R> connect(R&& recv);
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+    using operation_state_concept = operation_state_t;
+
+    template<class R>
+    @\exposid{state}@(coroutine_handle<promise_type> h, R&& rr);
```
```suggestion
      @\exposid{state}@(coroutine_handle<promise_type> h, R&& rr);
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\begin{itemdescr}
+\pnum
+The type \exposid{own-env-t} is \tcode{Environment::template
+env_type<decltype(get_env(\linebreak declval<Rcvr>()))>} if that % avoid Overfull
```
```suggestion
env_type<decltype(get_env(\brk{}declval<Rcvr>()))>} if that
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\pnum
+The type \exposid{own-env-t} is \tcode{Environment::template
+env_type<decltype(get_env(\linebreak declval<Rcvr>()))>} if that % avoid Overfull
+qualified-id is valid and denotes a type, \tcode{env<>} otherwise.
```
```suggestion
\grammarterm{qualified-id} is valid and denotes a type, \tcode{env<>} otherwise.
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+
+\begin{itemdecl}
+template<class R>
+@\exposid{state}@(coroutine_handle<promise_type> h, R&& rr);
```
```suggestion
  @\exposid{state}@(coroutine_handle<promise_type> h, R&& rr);
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\effects
+Initializes
+
+\item \exposid{handle} with \tcode{std::move(h)};
+\item \exposid{rcvr} with \tcode{std::forward<R>(rr)};
+\item \exposid{own-env}
+    with \tcode{\exposid{own-env-t}(get_env(\exposid{rcvr}))} if that expression
+    is valid and \tcode{\exposid{own-env-t}()} otherwise.
+    If neither of these expressions is valid, the program is ill-formed.
+\item \exposid{environment} with \tcode{Environment(\exposid{own-env})} if that expression is
+    valid, otherwise \tcode{Environment(get_env(\exposid{rcvr}))}
+    if this expression is valid, otherwise \tcode{Environment()}.
+    If neither of these expressions is valid, the program is ill-formed.
```
There is no `itemize` environment surrounding these.

I'm surprised it even builds, but I guess that's a typical TeX moment.

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\item \exposid{handle} with \tcode{std::move(h)};
+\item \exposid{rcvr} with \tcode{std::forward<R>(rr)};
+\item \exposid{own-env}
+    with \tcode{\exposid{own-env-t}(get_env(\exposid{rcvr}))} if that expression
```
This indentation within bullets is wrong. Just get rid of it an make it all flat.

dietmarkuehl replied:
I have removed the leading indentation with [this commit](https://github.com/cplusplus/draft/pull/8032/commits/2ec36bb49f287d73cd17e8dcf9ebbb0f13d3ceed).

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\pnum
+\effects
+Equivalent to:
+
```
Wrong empty line.

Eisenwave replied:
Same issue in a number of other places, I won't make any more separate comments for this.

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\pnum
+\effects
+Let \tcode{prom} be the object \tcode{\exposid{handle}.promise()}.
+Associates \tcode{STATE(prom)}, \tcode{RCVR(prom)}, and \tcode{SCHED(prom)}
```
```suggestion
Associates
\tcode{\exposid{STATE}(prom)},
\tcode{\exposid{RCVR}(prom)}, and
\tcode{\exposid{SCHED}(prom)}
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\tcode{\libconcept{invocable}<Fn>} and
+\tcode{\libconcept{constructible_from}<Fn, Init>} are modeled,
+\tcode{stop_token_type::callback_type<Fn>} models
+\tcode{\exposid{stoppable-callback-for}<Fn,\linebreak stop_token_type, % avoid Overfull
```
```suggestion
\tcode{\exposconcept{stoppable-callback-for}<Fn,\brk{} stop_token_type, Init>}.
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+    using type = remove_cvref_t<E>;
+    type error;
+  };
+  template<class E>
+  with_error(E) -> with_error<E>;
```
```suggestion
  template<class E>
    with_error(E) -> with_error<E>;
```
In general, there should be two spaces of indentation after a *template-head*, except for class definitions. I won't make any more comments for this.

dietmarkuehl replied:
I applied the change with [this commit](https://github.com/cplusplus/draft/pull/8032/commits/70456ebe2b2681393bf3da89887c1c1e48356150).

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+    void uncaught_exception();
+    coroutine_handle<> unhandled_stopped();
+
+    void return_void();             // present only if is_void_v<T> is true;
+    template<class V>
+    void return_value(V&& value);   // present only if is_void_v<T> is false;
```
```suggestion
    void return_void();               // present only if \tcode{is_void_v<T>} is \tcode{true};
    template<class V>
      void return_value(V&& value);   // present only if \tcode{is_void_v<T>} is \tcode{false};
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+    allocator_type    @\exposidnc{alloc}@;  // \expos
+    stop_source_type  @\exposidnc{source}@; // \expos
+    stop_token_type   @\exposidnc{token}@;  // \expos
+    optional<T>       @\exposidnc{result}@; // \expos; present only if is_void_v<T> is false;
```
```suggestion
    optional<T>       @\exposidnc{result}@; // \expos; present only if \tcode{is_void_v<T>} is \tcode{false};
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\pnum
+Let \tcode{prom} be an object of \tcode{promise_type} and let \tcode{tsk} be
+the \tcode{task} object created by \tcode{prom.get_return_object()}. The
+description below refers to objects \tcode{STATE(prom)},
+\tcode{RCVR(prom)}, and \tcode{SCHED(prom)} associated
```
Some more missing `\exposid` here.

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\pnum
+\returns
+A \tcode{task} object whose member \exposid{handle} is
+\tcode{coroutine_handle<promise_type>::\linebreak from_promise(*this)}. % avoid Overfull
```
```suggestion
\tcode{coroutine_handle<promise_type>::\brk{}from_promise(*this)}.
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\begin{itemdescr}
+\pnum
+\returns
+An awaitable object of unspecified type \iref{expr.await} whose
```
```suggestion
An awaitable object of unspecified type\iref{expr.await} whose
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\begin{itemdescr}
+\pnum
+\returns
+An awaitable object of unspecified type \iref{expr.await} whose
```
```suggestion
An awaitable object of unspecified type\iref{expr.await} whose
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+operation associated with \tcode{STATE(*this)} by invoking:
+
+\begin{itemize}
+\item
+    \tcode{set_error(std::move(RCVR(*this)), std::move(e))}
+    if \tcode{\exposid{errors}.index()} is greater than zero and
+    \tcode{e} is the value held by \exposid{errors}, otherwise
+\item
+    \tcode{set_value(std::move(RCVR(*this)))} if \tcode{is_void<T>} is \tcode{true},
+    and otherwise
+\item
+ \tcode{set_value(std::move(RCVR(*this)), *\exposid{result})}.
```
Remove indentation for bullets here.

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+
+\pnum
+\returns
+An awaitable object of unspecified type \iref{expr.await} whose
```
```suggestion
An awaitable object of unspecified type\iref{expr.await} whose
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\begin{itemdescr}
+\pnum
+\returns
+If \tcode{same_as<inline_scheduler, scheduler_type>} is \tcode{true}
```
```suggestion
If \tcode{@\libconcept{same_as}@<inline_scheduler, scheduler_type>} is \tcode{true}
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\pnum
+\returns
+If \tcode{same_as<inline_scheduler, scheduler_type>} is \tcode{true}
+returns \tcode{as_awaitable(\linebreak std::forward <Sender>(sndr), % avoid Overfull
+*this)}; otherwise returns
```
```suggestion
returns \tcode{as_awaitable(\brk{}std::forward<Sender>(sndr), *this)};
otherwise returns
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+If \tcode{same_as<inline_scheduler, scheduler_type>} is \tcode{true}
+returns \tcode{as_awaitable(\linebreak std::forward <Sender>(sndr), % avoid Overfull
+*this)}; otherwise returns
+\tcode{as_awaitable(affine_on(\linebreak std::forward<Sender>(sndr), % avoid Overfull
+SCHED(*this)), *this)}.
```
```suggestion
\tcode{as_awaitable(affine_on(\brk{}std::forward<Sender>(sndr), \exposid{SCHED}(*this)), *this)}.
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\begin{itemdescr}
+\pnum
+\effects
+Equivalent to: \tcode{returns await_transform(just(exchange(SCHED(*this),
+\linebreak scheduler_type(sch.scheduler))), *this);} % avoid Overfull
```
```suggestion
Equivalent to:
\begin{codeblock}
return await_transform(just(exchange(SCHED(*this),  scheduler_type(sch.scheduler))), *this);
\end{codeblock}
```
Note the typo in `returns` (paper has this), but I think this doesn't really need a separate commit.

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\pnum
+\effects
+If the signature \tcode{set_error_t(exception_ptr)} is not an element
+of \tcode{error_types}, calls \tcode{terminate()} \iref{except.terminate}.
```
```suggestion
of \tcode{error_types}, calls \tcode{terminate()}\iref{except.terminate}.
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\tcode{alloc} be \tcode{Allocator()}.  Let \tcode{arg_next} be the
+parameter following the first \tcode{allocator_arg_t} parameter (if
+any) and let \tcode{alloc} be \tcode{Allocator(arg_next)}.  Then
+\tcode{PAlloc} is \tcode{allocator_traits<Allocator>::template
+rebind_alloc\linebreak<U>} where \tcode{U} is an unspecified type % avoid Overfull
```
```suggestion
\tcode{PAlloc} is \tcode{allocator_traits<Allocator>::template
re\-bind_alloc\brk{}<U>} where \tcode{U} is an unspecified type
```

Eisenwave commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+coroutine state of size \tcode{size}, and unspecified additional
+state necessary to ensure that \tcode{operator delete} can later
+deallocate this memory block with an allocator equal to \tcode{palloc}.
+\pnum
```
```suggestion

\pnum
```

jensmaurer replied:
Yes, pnum needs an empty line in front (or a \begin{itemdescr}, otherwise it's not functioning properly.

Eisenwave reviewed (dismissed):
Thanks for drafting this up.

Lots of work required before this is ready to go.

Also note that fixing up CI builds is not a fixup commit; it should just go into the main commit. The point of fixups is just to distinguish between turning the paper into TeX, and fixing typos and other minor issues afterwards.

Massaging whitespace or paragraph justification is not a separate commit.

jensmaurer commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+where \tcode{CS} is \tcode{completion_signatures<set_value_t()>},
+then:
+
+\begin{itemize}
+\item the expression \tcode{connect(sndr, rcvr)} has
+type \tcode{\exposid{inline-state}<remove_cvref_t<decltype((rcvr))>>}
+and is potentially-throwing if and only if
+\tcode{((void)sndr, auto(rcvr))} is potentially-throwing, and
+\item the expression
+\tcode{get_completion_scheduler<set_value_t>(get_env(sndr))} has
+type\linebreak\tcode{inline_scheduler} and is potentially-throwing % avoid Overfull
+if and only if \tcode{get_env(sndr)} is potentially-throwing.
+\end{itemize}
```
I think other parts of senders-receivers do sndr and rcvr as code font; I'd be ok with trying to be consistent here.  But please check my assumption.

dietmarkuehl replied:
Based on looking at other sections of `exec.tex` I believe Jens is right: the function parameters use neither `\placeholder` nor `\exposid`. I'll leave it consistent unless told otherwise.

Eisenwave commented on source/exec.tex:
```diff
@@ -708,6 +708,24 @@
+  class @\libglobal{task_scheduler}@;
+
+  // \ref{exec.task}
+  template<class E>
```
As dietmar explained above, those definitions exist, but are missing declarations in the synopsis.

jensmaurer replied:
Ah, ok.  And those declarations seem to have no add-on info, so they probably should exist in the synopsis only.

dietmarkuehl replied:
I have created a commit moving the definition of these two class templates to the synopsis (it is a [separate commit](https://github.com/cplusplus/draft/pull/8032/commits/1c45cffe76d86f395744ea12072e55a1e37e97db)).

dietmarkuehl commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\begin{codeblock}
+namespace std::execution {
+  class task_scheduler {
+    class @\exposidnc{ts-sender}@; // \expos
```
The CI over here complained that `sender` should reference the `sender`. However, `sender` is a local class name (which does model `sender`), i.e., it isn't a reference to the `sender` concept. If there is a different work-around which keeps the name `sender` for the exposition-only type I'm happy to use it.

Eisenwave replied:
Ok weird stuff, that's beyond my TeX expertise.

dietmarkuehl commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+    class @\exposidnc{state}@;  // \expos
+
+  public:
+    using scheduler_concept = scheduler_t;
```
You mean I should use something like `@\libglobal{scheduler_concept}@`? In other place in this section (mostly using `using sender_concept = sender_t;`) there isn't any additional annotation. `scheduler_concept` is a naming coming from the [`scheduler` concept](https://eel.is/c++draft/exec.sched).

dietmarkuehl commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+except that \tcode{sndr} is evalutated only once.
+
+\pnum
+The exposition-only class template \exposid{impls-for}\iref{exec.snd.general}
```
There were just two \irefs which didn't match references in the original text. I have removed them (with [this commit](https://github.com/cplusplus/draft/pull/8032/commits/5df4a38d14d138778b3a80307410f96f8c34af13)).

dietmarkuehl commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\tcode{((void)sndr, auto(rcvr))} is potentially-throwing, and
+\item the expression
+\tcode{get_completion_scheduler<set_value_t>(get_env(sndr))} has
+type\linebreak\tcode{inline_scheduler} and is potentially-throwing % avoid Overfull
```
I have remove all uses of `\linebreak` and addressed all overfull hboxes.

dietmarkuehl commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+\end{itemdescr}
+
+\begin{itemdecl}
+@\exposid{ts-sender}@ schedule();
```
I have addressed the missing index members with [this commit](https://github.com/cplusplus/draft/pull/8032/commits/9953567985c8c3720b2d0561ca8802505684f1a7).

dietmarkuehl commented on source/exec.tex:
```diff
@@ -5672,3 +5690,845 @@
+
+\begin{itemize}
+\item \tcode{allocator_type} is \tcode{Environment::allocator_type}
+if that qualified-id is valid and denotes a type, \tcode{allocator<byte>}
```
I have wrapped `qualified_id` int `\grammarterm` with [this commit](https://github.com/cplusplus/draft/pull/8032/commits/619f2ad92587408001f3d42db0dc1f3b66efcd1e).

dietmarkuehl commented:
I believe I have addressed all comments which were raised so far. Please let me know when I need to do more changes.

Thank you very much for the review!

tkoeppe commented on source/exec.tex:
```diff
@@ -7877,3 +7909,871 @@
 as temporary storage for the duration of the operation launched by this call.
 \end{itemdescr}
+
+\rSec2[exec.affine.on]{\tcode{execution::affine_on}}
```
@dietmarkuehl I'm assuming this material is meant to go under "Coroutine utilities" (which happened to be at the end of [exec] prior to this meeting), rather than under "Namespace system_context_replaceability".

dietmarkuehl replied:
yes, that is correct.

tkoeppe replied:
No, I've got it. It was just that the edit instructions were not literally true ("at the end of [exec]" should instead of named the subclause at whose end you wanted it), so I mainly just want to confirm that I understood the intention.

tkoeppe commented:
Thanks a lot for this, Dietmar! Very nice drafting!
